eos-tech-support/eos-write-live

```diff
@@ -0,0 +1,272 @@
+    exit 1
+}
+
+command -v unzip >/dev/null || {
+    echo "unzip is not installed... aborting!" >&2
+    echo "Try 'sudo apt-get install unzip'" >&2
+    exit 1
+}
+
+if test \! -b $DEVICE; then
+    echo "$DEVICE does not exist or is not a block device... aborting!" >&2
+    exit 1
+fi
+
+if grep -qs $DEVICE /proc/mounts; then
+    # Protect against overwriting the device currently in use
+    echo "$DEVICE is currently in use -- please unmount and try again" >&2
+    exit 1
+fi
+
+if ! $FORCE; then
+    read -p "Are you sure you want to overwrite all data on $DEVICE? [y/N] "
+    response=${REPLY,,} # to lower
+    if [[ ! $response =~ ^(yes|y)$ ]]; then
+        exit 1
+    fi
+fi
+
+if [ ! -z "$FETCH_LATEST" ]; then
+    if [ -z "$OS_IMAGE" ]; then
+        OS_IMAGE=$($EOS_DOWNLOAD_IMAGE --product eos)
+    fi
+    if [ -z "$WINDOWS_TOOL" ]; then
+        WINDOWS_TOOL=$($EOS_DOWNLOAD_IMAGE --windows-tool)
+    fi
+fi
+
+if [ ! -f "$WINDOWS_TOOL" -o ! -f "$OS_IMAGE" ]; then
+    echo "--os-image and --windows-tool are required if --latest is not specified" >&2
+    usage >&2
+    exit 2
+fi
+
+check_exists "$OS_IMAGE" "image"
+
+EXTRACTED_SIGNATURE="${OS_IMAGE%.gz}.asc"
+check_exists "$EXTRACTED_SIGNATURE" "uncompressed image signature"
+
+BOOT_ZIP="${OS_IMAGE%.img.gz}.boot.zip"
+check_exists "$BOOT_ZIP" "bootloader bundle"
+check_exists "$BOOT_ZIP.asc" "bootloader bundle signature"
+
+check_exists "$WINDOWS_TOOL" "Windows tool"
+
+echo
+echo "Partitioning $DEVICE..."
+
+DEVICE_IMAGES="${DEVICE}1"
+if $MBR; then
+    (
+        echo "1 : type=07"
+    ) | sfdisk --label dos $DEVICE
+else
+    # https://en.wikipedia.org/wiki/GUID_Partition_Table#Partition_type_GUIDs
+    PARTITION_SYSTEM_GUID="c12a7328-f81f-11d2-ba4b-00a0c93ec93b"
+    PARTITION_BIOS_BOOT_GUID="21686148-6449-6E6F-744E-656564454649"
+    PARTITION_BASIC_DATA_GUID="ebd0a0a2-b9e5-4433-87c0-68b6b72699c7"
+
+    # We want the data partition, "eosimages", to occupy all the space on the disk
+    # after the UEFI and BIOS boot partitions. But, we also want it to be numbered
+    # first: apparently Windows will only mount the partition numbered first,
+    # regardless of where it is on the disk.
+    DEVICE_EFI="${DEVICE}2"
+    DEVICE_BIOS="${DEVICE}3"
+    (
+        echo "2 : start=2048, size=62MiB, type=$PARTITION_SYSTEM_GUID"
+        echo "3 : size=1MiB, type=$PARTITION_BIOS_BOOT_GUID"
+        echo "1 : name=eosimages, type=$PARTITION_BASIC_DATA_GUID"
+    ) | sfdisk --label gpt $DEVICE
+fi
+
+partprobe
+udevadm settle
+
+echo
+echo "Creating partitions"
+if ! $MBR; then
+    mkfs.vfat -n efi "${DEVICE_EFI}"
+fi
+$MKFS_IMAGES $MKFS_ARGS eosimages "${DEVICE_IMAGES}"
+
+# Give udisks a chance to notice the new partitions
+partprobe
+udevadm settle
+
+udisks_mount() {
+    udisksctl mount -b "$1" --no-user-interaction >&2
+    MOUNT_POINT=$(udisksctl info -b "$1" | grep "MountPoints:" | cut -f2 -d':')
+    test -d "$MOUNTPOINT"
+    echo $MOUNT_POINT
+}
```

Comment:
I might be wrong, but I actually do think that if the device is a USB stick, udisks will try to automount it from a graphical session once valid filesystems show up. I've definitely seen QA struggle with that and use stuff like `sudo umount -a` to try to clean up with hilarious side effects.

I think it's probably fine as is, but one way to handle this would be to preemptively unmount:
```
udisksctl unmount -b "$1" --no-user-interaction 2>/dev/null || true
```